src/javascript/function-scanner.ts

```diff
@@ -0,0 +1,157 @@
+    const visitor = new FunctionVisitor(document);
+    await visit(visitor);
+    return Array.from(visitor.functions);
+  }
+}
+
+class FunctionVisitor implements Visitor {
+  functions = new Set<ScannedFunction>();
+  document: JavaScriptDocument;
+
+  constructor(document: JavaScriptDocument) {
+    this.document = document;
+  }
+
+  /**
+   * Scan standalone function declarations.
+   */
+  enterFunctionDeclaration(
+      node: estree.FunctionDeclaration, _parent: estree.Node) {
+    this._initFunction(node, getIdentifierName(node.id), node);
+    return;
+  }
+
+  /**
+   * Scan functions assigned to newly declared variables.
+   */
+  enterVariableDeclaration(
+      node: estree.VariableDeclaration, _parent: estree.Node) {
+    if (node.declarations.length !== 1) {
+      return;  // Ambiguous.
+    }
+    const declaration = node.declarations[0];
+    const declarationId = declaration.id;
+    const declarationValue = declaration.init;
+    if (declarationValue && isFunctionType(declarationValue)) {
+      return this._initFunction(
+          node, objectKeyToString(declarationId), declarationValue);
+    }
+  }
+
+  /**
+   * Scan functions assigned to variables and object properties.
+   */
+  enterAssignmentExpression(
+      node: estree.AssignmentExpression, parent: estree.Node) {
+    if (isFunctionType(node.right)) {
+      this._initFunction(parent, objectKeyToString(node.left), node.right);
+    }
+  }
+
+  /**
+   * Scan functions defined inside of object literals.
+   */
+  enterObjectExpression(node: estree.ObjectExpression, _parent: estree.Node) {
+    for (let i = 0; i < node.properties.length; i++) {
+      const prop = node.properties[i];
+      const propValue = prop.value;
+      const name = objectKeyToString(prop.key);
+      if (isFunctionType(propValue)) {
+        this._initFunction(prop, name, propValue);
+        continue;
+      }
+      const comment = getAttachedComment(prop) || '';
+      const docs = jsdoc.parseJsdoc(comment);
+      if (jsdoc.getTag(docs, 'function')) {
+        this._initFunction(prop, name);
+        continue;
+      }
+    }
+  }
+
+  private _initFunction(
+      node: estree.Node, anaylzedName?: string, _fn?: estree.Function) {
+    const comment = getAttachedComment(node);
+
+    // Quickly filter down to potential candidates.
+    if (!comment || comment.indexOf('@memberof') === -1) {
+      return;
+    }
+
+    if (!anaylzedName) {
+      // TODO(fks): Propagate a warning if name could not be determined
+      return;
+    }
+
+    const docs = jsdoc.parseJsdoc(comment);
+    const functionName = getNamespacedIdentifier(anaylzedName, docs);
+    const sourceRange = this.document.sourceRangeForNode(node)!;
+    const returnTag = jsdoc.getTag(docs, 'return');
+    const summaryTag = jsdoc.getTag(docs, 'summary');
+    const summary = (summaryTag && summaryTag.description) || '';
+    const description = docs.description;
+
+    let functionReturn;
```

Comment:
Woah, clang-format does not like that:

```

    const functionReturn = (returnTag) ? {
      type: returnTag.type || undefined,
      desc: returnTag.description || '',
    } :
                                         undefined;
```

I'm going to stick with let, I agree const is great but not at the expense of significant clarity.